torch/_inductor/codegen/cpp_micro_gemm.py

```diff
@@ -504,18 +504,67 @@ class CppMicroGemmAMX(CppMicroGemm):
     This class generates the code for micro gemm using Advanced Matrix eXtention (AMX)
     instructions available in 4th generation Intel Xeon for compute.
     It supports input types of torch.bfloat16 with fp32 output.
-    TODO(jgong5): support int8 data type.
     """
 
     TEMPLATE_ENTRY = r"""
 {{declare_kernel}} {
     {{kernel.assert_function}}(N % {{block_n}} == 0, "N dimension must be multiple of {{block_n}}");
     {{kernel.assert_function}}(K % 2 == 0, "K dimension must be multiple of 2");
+{%- if input_dtype == torch.bfloat16 and input2_dtype == torch.int8 %}
+    // Create a stack-allocated buffer for tiles of B.
+    // Except maybe for the tail-case, an AMX tile of B has 16x32 BF16 elements
+    // It's possible that N may not be equal to Nr, but might be a multiple of it.
+    // In that case, Kc (which is K in the microkernel) would be smaller (than the case in which N == Nr).
+    const auto num_elements_per_b_tile = 512;
+    const auto buf_size_per_nr_block = ((K + {{block_k}} - 1) / {{block_k}}) * num_elements_per_b_tile * 2;
+    const auto buf_size = buf_size_per_nr_block * (N / {{block_n}});
+    alignas(4096) {{input_t}} bf16_weights_buf[buf_size];
+
+    const auto last_k_offset = K / {{block_k}} * {{block_k}};
+    const auto tail_k_size = K - last_k_offset;
+    const auto num_b_rows = (last_k_offset > 0) ? 16 : (tail_k_size * sizeof({{input_t}})) / 4;
+    const auto b_tile_ptr_stride = ldb * {{vnni_size}};
+
+    auto load_B_row = [&]({{input2_t}}* {{restrict_keyword}} src, {{input_t}}* {{restrict_keyword}} dst) {
+        auto b_int8 = at::vec::Vectorized<int8_t>::loadu(src, static_cast<int64_t>(32));
+        auto b_bf16 = at::vec::convert<{{input_t}}>(b_int8);
+        b_bf16.store(dst);
+    };
+
+    auto load_B_in_buf = [&]({{input2_t}}* B_ptr, int idx) {
+        {{input_t}}* base_addr = bf16_weights_buf + idx;
+        {{kernel.unroll_pragma(8)}}
+        for (int i = 0; i < num_b_rows; i++) {
+            load_B_row(
+                B_ptr + i * b_tile_ptr_stride,
+                base_addr + i * 32
+            );
+        }
+    };
+    auto load_weights = [&](int n) {
+        // Load weights & cache them in L1D.
+        const int64_t init_idx =  n * buf_size_per_nr_block;
+        {{kernel.unroll_pragma(4)}}
+        for (int k = 0; k < last_k_offset; k += {{block_k}}) {
```

Comment:
May I know how the tail of K is loaded?

Reply:
To make things simple, we'll allocate space for a whole 16x32 tile even for the tail-case. This is being done even if `N != Nr`

Reply:
I don't quite understand the logic here - k loops over 0 to `last_k_offset` but not over the tail, right?

Reply:
Thanks! You're right! The existing code actually computes for the tail separately

https://github.com/pytorch/pytorch/blob/main/torch/_inductor/codegen/cpp_micro_gemm.py#L656-L678

So, I'll change the logic to load weights.